sigstore/_internal/rekor/_client.py

```diff
@@ -2,8 +2,95 @@
 Client implementation for interacting with Rekor.
 """
 
+import json
+from abc import ABC
+from typing import Any, Optional, Tuple
+from urllib.parse import urljoin
+
+import requests  # type: ignore
+
+DEFAULT_REKOR_URL = "https://rekor.sigstore.dev/api/v1/"
+
+
+class Endpoint(ABC):
+    def __init__(self, url: str, session: requests.Session) -> None:
+        self.url = url
+        self.session = session
+
 
 class RekorClient:
     """The internal Rekor client"""
 
+    def __init__(self, url: str = DEFAULT_REKOR_URL) -> None:
+        self.url = url
+        self.session = requests.Session()
+        self.session.headers.update(
+            {"Content-Type": "application/json", "Accept": "application/json"}
+        )
+
+    @property
+    def index(self) -> Endpoint:
+        return RekorIndex(urljoin(self.url, "index/"), session=self.session)
+
+    @property
+    def log(self) -> Endpoint:
+        return RekorLog(urljoin(self.url, "log/"), session=self.session)
+
+
+class RekorIndex(Endpoint):
+    @property
+    def retrieve(self) -> Endpoint:
+        return RekorRetrieve(urljoin(self.url, "retrieve/"), session=self.session)
+
+
+class RekorRetrieve(Endpoint):
+    def post(self, sha256_hash: Optional[str] = None) -> requests.Response:
+        data = {"hash": f"sha256:{sha256_hash}"}
```

Comment:
The `sha256:` prefix shouldn't be necessary according to the OpenAPI spec; I opened an issue when I ran into problems without it [here](https://github.com/sigstore/rekor/issues/699). Since it won't work without the prefix as of now, this is fine. But just a note for the future.

Reply:
Yeah, unfortunately I'm not sure how much we can trust the OpenAPI spec to be correct and/or exist. It was removed for Fulcio entirely (https://github.com/sigstore/fulcio/pull/262). It probably makes the most sense for us to hand-craft our clients using the spec as a reference, but check them for accuracy against the live API.